iceoryx2-bb/log/src/logger/file.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! # Example
+//!
+//! Using the file logger.
+//!
+//! ```no_run
+//! use iceoryx2_bb_log::{info, set_logger, set_log_level, LogLevel, logger::file};
+//! use std::sync::LazyLock;
+//!
+//! const LOG_FILE: &str = "fuu.log";
+//! static FILE_LOGGER: LazyLock<file::Logger> = LazyLock::new(|| file::Logger::new(LOG_FILE));
+//! set_logger(&*FILE_LOGGER);
+//! set_log_level(LogLevel::Trace);
+//!
+//! // written into log file "fuu.log"
+//! info!("hello world");
+//! ```
+
+// TODO: [Reminder to my future self]
+// In the long-term the file logger may be required to be based on the same
+// iceoryx2_pal_posix platform. In this case, the logger needs to use the low-level calls directly
+// to avoid a circular dependency with iceoryx2_bb_posix.
+use std::{
+    collections::VecDeque,
+    fmt::Debug,
+    fs::OpenOptions,
+    io::Write,
+    sync::{Arc, Condvar, Mutex},
+    thread::JoinHandle,
+    time::{Duration, Instant, SystemTime},
+};
+
+use crate::LogLevel;
+
+struct Entry {
+    timestamp: Duration,
+    elapsed_time: Duration,
+    log_level: LogLevel,
+    origin: String,
+    message: String,
+}
+
+impl Debug for Entry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "timestamp: {:?}, elapsed_time: {:?}, log_level: {:?}, origin: {}, message: {}",
+            self.timestamp, self.elapsed_time, self.log_level, self.origin, self.message
+        )
+    }
+}
+
+struct State {
+    buffer: VecDeque<Entry>,
+    keep_running: bool,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            buffer: VecDeque::new(),
+            keep_running: true,
+        }
+    }
+}
+
+/// A logger that logs all messages into a file. It implements an active object pattern. A
+/// background thread waits on a queue of log messages and whenever a new message is added.
+pub struct Logger {
+    state: Arc<Mutex<State>>,
+    trigger: Arc<Condvar>,
+    start_time: Instant,
+    _background_thread: Arc<Option<JoinHandle<()>>>,
+}
+
+impl Logger {
```

Comment:
Not really an expert on loggers, but the approach makes sense to me.

The use of a separate thread raised some question marks, though. It adds some jitter, but the tradeoff would add latency to the user thread. Fine for me. Can change if more information later reveals another approach.